File: apps/android/core/common/src/main/kotlin/app/gauja/core/common/AppError.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.common

enum class AppError(val messageKey: String) {
    NETWORK("error.network"),
    TLS("error.tls"),
    AUTH("error.auth"),
    PERMISSION("error.permission"),
    NOT_FOUND("error.notFound"),
    VALIDATION("error.validation"),
    SERVER_VERSION("error.serverVersion"),
    OFFLINE("error.offline"),
    UNKNOWN("error.unknown"),
}

// Exceptions carry only an allow-listed category, never an underlying HTTP body or credential.
class AppException(val reason: AppError) : Exception(reason.messageKey)

sealed interface AppResult<out T> {
    data class Success<T>(val value: T) : AppResult<T>

    data class Failure(val error: AppError) : AppResult<Nothing>
}

File: apps/ios/Packages/Common/Sources/Common/AppError.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
public enum AppError: String, Error, Sendable {
    case network, tls, auth, permission, notFound, validation, serverVersion, offline, unknown
    public var messageKey: String { "error.\(rawValue)" }
}

public typealias AppResult<Value> = Result<Value, AppError>
